package/MDAnalysis/analysis/dssp/dssp.py

```diff
@@ -0,0 +1,364 @@
+        positions on PRO residues, and an N-terminal residue.
+
+    Raises
+    ------
+    ValueError
+        if ``guess_hydrogens`` is True but some non-PRO hydrogens are missing.
+
+    Examples
+    --------
+
+    For example, you can assign secondary structure for a single PDB file:
+
+    >>> from MDAnalysis.analysis.dssp import DSSP
+    >>> from MDAnalysisTests.datafiles import PDB
+    >>> import MDAnalysis as mda
+    >>> u = mda.Universe(PDB)
+    >>> run = DSSP(u).run()
+    >>> print("".join(run.results.dssp[0][:20]))
+    --EEEEE-----HHHHHHHH
+
+    Also, per-frame dssp assignment allows you to build average
+    secondary structure -- ``DSSP.results.dssp_ndarray`` holds
+    (n_frames, n_residues, 3) shape ndarray with one-hot encoding of
+    loop, helix and sheet, respectively:
+
+    >>> from MDAnalysis.analysis.dssp import translate, DSSP
+    >>> from MDAnalysisTests.datafiles import TPR, XTC
+    >>> u = mda.Universe(TPR, XTC)
+    >>> long_run = DSSP(u).run()
+    >>> mean_secondary_structure = translate(long_run.results.dssp_ndarray.mean(axis=0))
+    >>> print(''.join(mean_secondary_structure)[:20])
+    -EEEEEE------HHHHHHH
+
+    .. versionadded:: 2.8.0
+    """
+
+    def __init__(self, atoms: Union[Universe, AtomGroup], guess_hydrogens: bool = True):
+        self._guess_hydrogens = guess_hydrogens
+
+        ag: AtomGroup = (
+            atoms.select_atoms("protein") if isinstance(atoms, Universe) else atoms
+        )
+        super().__init__(ag.universe.trajectory)
+
+        # define necessary selections
+        # O1/OT1 is for C-terminal residue
+        heavyatom_names = ("N", "CA", "C", "O O1 OT1")
+        self._heavy_atoms: dict[str, "AtomGroup"] = {
+            t: ag.select_atoms(f"protein and name {t}") for t in heavyatom_names
+        }
+        self._hydrogens: list["AtomGroup"] = [
+            res.atoms.select_atoms("name H")
+            for res in ag.select_atoms("protein").residues
+        ]
+        # can't do it the other way because I need missing values to exist
+        # so that I could fill them in later
+
+        if not self._guess_hydrogens:
+            for calpha, hydrogen in zip(
+                self._heavy_atoms["CA"][1:], self._hydrogens[1:]
+            ):
+                if not hydrogen and calpha.resname != "PRO":
+                    raise ValueError(
+                        (
+                            "Universe is missing non-PRO hydrogen "
+                            f"on residue {calpha.residue}"
+                        )
+                    )
+
+    def _prepare(self):
+        self.results.dssp_ndarray = []
+
+        positions = [group.positions for group in self._heavy_atoms.values()]
+        if len(set(map(lambda arr: arr.shape[0], positions))) != 1:
+            raise ValueError(
+                (
+                    "Universe contains not equal number of (N,CA,C,O) atoms ('name' field)."
```

Comment:
```suggestion
                    "Universe contains unequal numbers of (N,CA,C,O) atoms ('name' field)."
```
Shouldn't this check be done on `__init__` after the heavyatom selection step?

Reply:
Indeed, changed.